cpp/autosar/src/rules/A12-8-6/CopyAndMoveNotDeclaredProtected.ql

```diff
@@ -20,31 +20,45 @@ predicate isInvalidConstructor(Constructor f, string constructorType) {
   not f.isDeleted() and
   not f.isProtected() and
   (
-    f instanceof MoveConstructor and constructorType = "Move constructor"
+    f instanceof MoveConstructor and
+    if f.isCompilerGenerated()
+    then constructorType = "Implicit move constructor"
+    else constructorType = "Move constructor"
     or
-    f instanceof CopyConstructor and constructorType = "Copy constructor"
+    f instanceof CopyConstructor and
+    if f.isCompilerGenerated()
+    then constructorType = "Implicit copy constructor"
+    else constructorType = "Copy constructor"
   )
 }
 
 predicate isInvalidAssignment(Operator f, string operatorType) {
   not f.isDeleted() and
   (
-    f instanceof CopyAssignmentOperator and operatorType = "Copy assignment operator"
+    f instanceof MoveAssignmentOperator and
+    if f.isCompilerGenerated()
+    then operatorType = "Implicit move assignment operator"
+    else operatorType = "Move constructor"
     or
-    f instanceof MoveAssignmentOperator and operatorType = "Move assignment operator"
+    f instanceof CopyAssignmentOperator and
+    if f.isCompilerGenerated()
+    then operatorType = "Implicit copy assignment operator"
```

Comment:
I like this addition to differentiate implicit vs user defined, 

quick check, does it make sense to customize the alert a bit more even?
right now we have some alerts like `Implicit copy assignment operator for base class 'Message' is not declared protected or deleted.`

but it doesnt make sense to "declare an implicit copy constructor as protected". it would make sense to say that the implicit one isnt deleted though... open to suggestions!

Reply:
That is a good suggestion. I will make the alert more precise.